### PR TITLE
Add helper methods to DBusValue to reduce casting.

### PR DESCRIPTION
**Pull request by robert-ancell**
The intention was to make helper methods to reduce the amount of typing and brackets required when using casting. However, I wanted to make it clear that it still is casting internally, which is why the methods are all prefixed with 'as'.

Not sure if the documentation on the methods is the best, but found it difficult to come up with better words than 'Converts' and 'Extracts'. Some types are just a cast because this can't be natively represented, while others are just simple types.

### CODE REVIEW

File: lib/src/dbus_value.dart [hunk 1]
```diff
@@ -37,6 +37,102 @@ abstract class DBusValue {
...
+  DBusObjectPath asObjectPath() => this as DBusObjectPath;
+
+  /// Extracts the signature inside this value. Only works if [signature] is 'g'.
+  DBusSignature asSignature() => this as DBusSignature;
+
+  /// Extracts the value stored inside this variant. Only works if [signature] is 'v'.
+  DBusValue asVariant() => (this as DBusVariant).value;
+
+  /// Extracts the maybe type inside this value. Only works if [signature] is a maybe type, e.g. 'mi'.
+  DBusMaybe asMaybe() => this as DBusMaybe;
```
Comment (jpnurmi) [inline lib/src/dbus_value.dart]:
```suggestion
  DBusValue? asMaybe() => (this as DBusMaybe).value;
```

File: lib/src/dbus_value.dart [hunk 2]
```diff
@@ -37,6 +37,102 @@ abstract class DBusValue {
...
+
+  /// Extracts the [ResourceHandle] inside this unix file descriptor D-Bus value. Only works if [signature] is 'h'.
+  ResourceHandle asUnixFd() => (this as DBusUnixFd).handle;
+
+  /// Extracts this child values inside this struct. Only works if [signature] is a struct type, e.g '(si)'.
+  Iterable<DBusValue> asStruct() => (this as DBusStruct).children;
+
+  /// Extracts the array inside this value. Only works if [signature] is an array type, e.g 'as'.
+  DBusArray asArray() => this as DBusArray;
```
Comment (jpnurmi) [inline lib/src/dbus_value.dart]:
```suggestion
  List<DBusValue> asArray() => (this as DBusArray).children;
```

File: lib/src/dbus_value.dart [hunk 3]
```diff
@@ -37,6 +37,102 @@ abstract class DBusValue {
...
+
+  /// Extracts the signatures inside this array. Only works if [signature] is 'ag'.
+  Iterable<DBusSignature> asSignatureArray() =>
+      (this as DBusArray).mapSignature();
+
+  /// Extracts the values inside this variant array. Only works if [signature] is 'av'.
+  Iterable<DBusValue> asVariantArray() => (this as DBusArray).mapVariant();
+
+  /// Extracts the dictionary inside this vlaue. Only works if [signature] is a dictionary type, e.g 'a{os}'.
+  DBusDict asDict() => this as DBusDict;
```
Comment (jpnurmi) [inline lib/src/dbus_value.dart]:
```suggestion
  Map<DBusValue, DBusValue> asDict() => (this as DBusDict).children;
```

**Review by jpnurmi — COMMENTED** [review]
Awesome! This makes a big difference in how convenient it is to read the values. :+1:

Given that the new API is for convenient reading of the contained values, I'd propose following the `(this as DBusType).value/children` pattern for `DBusMaybe`, `DBusArray`, and `DBusDict` too. The old syntax would be ok for the (rare?) cases when there's a need to access the containing DBus type.

**Comment by robert-ancell** [comment]
> Awesome! This makes a big difference in how convenient it is to read the values. +1
> 
> Given that the new API is for convenient reading of the contained values, I'd propose following the `(this as DBusType).value/children` pattern for `DBusMaybe`, `DBusArray`, and `DBusDict` too. The old syntax would be ok for the (rare?) cases when there's a need to access the containing DBus type.

I did it because I thought we would be losing useful information, but as you point out you can manually cast if necessary. Also the normal usage is to test the signature of the top-level object, which means all the sub-objects should have the signatures you expect.

**Pull request merged**